File: 0198-house-robber/0198-house-robber.java
class Solution {
    private int solve(int ind ,int arr[],int dp[]){
        if(ind == 0){
            return arr[ind];
        }
        if(ind < 0 ){
            return 0;
        }
        if(dp[ind]!= -1){
            return dp[ind];
        }
        int rob = arr[ind] + solve(ind - 2,arr,dp);
        int skip = 0 + solve(ind-1,arr,dp);
        return dp[ind] = Math.max(rob,skip);
    }
    public int rob(int[] nums) {
        int n = nums.length;
        int dp[] = new int[n];
        Arrays.fill(dp,-1);
        return solve(n-1,nums,dp);
    }
}

File: 0198-house-robber/0198-house-robber.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        int prev2 = 0;
        int prev = nums[0];
        for(int i = 0; i < nums.size(); i++) {
            int notTake = 0 + prev;
            int take = nums[i];
            if(i > 1) {
                take += prev2;
            }
            prev2 = prev;
            prev = max(take , notTake);
        }
        return prev;
    }
};